gateway-service/src/main/java/org/zowe/apiml/gateway/security/service/token/OIDCTokenProvider.java

```diff
@@ -12,111 +12,159 @@
+        HttpGet getRequest = new HttpGet(jwksUri + "?client_id=" + clientId);
         try {
-            CloseableHttpResponse response = httpClient.execute(post);
+            CloseableHttpResponse response = httpClient.execute(getRequest);
             final int statusCode = response.getStatusLine() != null ? response.getStatusLine().getStatusCode() : 0;
             final HttpEntity responseEntity = response.getEntity();
             String responseBody = "";
             if (responseEntity != null) {
                 responseBody = EntityUtils.toString(responseEntity, StandardCharsets.UTF_8);
             }
             if (statusCode == HttpStatus.SC_OK && !responseBody.isEmpty()) {
-                return mapper.readValue(responseBody, OIDCTokenClaims.class);
+                jwks.clear();
+                JwkKeys jwkKeys = mapper.readValue(responseBody, JwkKeys.class);
+                jwks.putAll(processKeys(jwkKeys));
             } else {
-                log.error("Failed to validate the OIDC access token. Unexpected response: {}", statusCode);
-                return null;
+                log.error("Failed to obtain JWKs from URI {}. Unexpected response: {}, response text: {}", jwksUri, statusCode, responseBody);
             }
-        } catch (IOException e) {
-            log.error("Failed to validate the OIDC access token. ", e);
+        } catch (IOException | IllegalStateException e) {
+            log.error("Error processing response from URI {}", jwksUri, e.getMessage());
         }
-        return null;
     }
 
+    private Map<String, Key> processKeys(JwkKeys jwkKeys) {
+        return jwkKeys.getKeys().stream()
+            .filter(jwkKey -> "sig".equals(jwkKey.getUse()))
+            .filter(jwkKey -> "RSA".equals(jwkKey.getKty()))
+            .collect(Collectors.toMap(JwkKeys.Key::getKid, jwkKey -> {
+                BigInteger modulus = base64ToBigInteger(jwkKey.getN());
+                BigInteger exponent = base64ToBigInteger(jwkKey.getE());
+                RSAPublicKeySpec rsaPublicKeySpec = new RSAPublicKeySpec(modulus, exponent);
+                try {
+                    KeyFactory keyFactory = KeyFactory.getInstance("RSA");
+                    return keyFactory.generatePublic(rsaPublicKeySpec);
+                } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
+                    throw new IllegalStateException("Failed to parse public key");
+                }
+           }));
+    }
+
+    private BigInteger base64ToBigInteger(String value) {
+        return new BigInteger(1, Decoders.BASE64URL.decode(value));
+    }
+
+    @Override
+    public boolean isValid(String token) {
+        if (StringUtils.isBlank(token)) {
+            log.debug("No token has been provided.");
+            return false;
+        }
+        Claims claims = null;
+        for (Map.Entry<String, Key> entry : jwks.entrySet()) {
+            claims = validate(token, entry.getValue());
```

Comment:
can we validate JWT only with JWK selected by "kid"?

Reply:
We need to parse the token to get the `kid` in the first place, can we parse it without passing in the key?

Reply:
I tried it and it works, I'll push it

Reply:
Sorry, it was misleading, the parsing fails if the key is not provided and I cannot get the header `kid` field of the token. I will have to go back to the previous implmentation

Reply:
I think we just need to remove signature before parsing JWT to get kid.

Reply:
it is probably better to just iterate over map entries as originaly propose, for now there should be only 1 JWK in the returned set anyway, we can adjust in the future

Reply:
there's actually two returned by our current provider's authorization server.
I tried as you said (removing the last segment) and it works.